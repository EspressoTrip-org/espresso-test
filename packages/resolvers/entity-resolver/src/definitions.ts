import * as cardinal from '@journeyapps-platform/types-cardinal';
import * as t from 'ts-codec';

export const Document = t
  .object({
    id: t.string,
    labels: t.record(t.string).optional()
  })
  // the design of this type is supposed to accept any additional keys,
  // AND this is valid in ts-codec when encoding and decoding, but this
  // specifically will cause validation issues when converting to json-schema (for validation)
  // in places where this type is used, we need to add {allowAdditional: true } in the codec validators
  // in order to get this to work correctly.
  .and(t.record(t.any));
export type Document = t.Encoded<typeof Document>;

export const EntityFilter = t.record(t.string.or(t.array(t.string)));
export type EntityFilter = t.Encoded<typeof EntityFilter>;

export const SpeculativeData = t.record(t.array(Document));
export type SpeculativeData = t.Encoded<typeof SpeculativeData>;

export const ResolveResourceIdParams = t.object({
  actions: t.array(t.string).optional(),
  model: t.string,
  policies: t.array(cardinal.RawPolicy),
  entity_filter: EntityFilter.optional(),
  speculative_data: SpeculativeData.optional()
});
export type ResolveResourceIdParams = t.Encoded<typeof ResolveResourceIdParams>;

export const CanAccessResourceParams = t
  .omit(ResolveResourceIdParams, ['entity_filter'])
  .and(t.object({ id: t.string }));
export type CanAccessResourceParams = t.Encoded<typeof CanAccessResourceParams>;

export const InvalidateParams = t.object({
  model: t.string,
  ids: t.array(t.string).optional()
});
export type InvalidateParams = t.Encoded<typeof InvalidateParams>;

//---- Interfaces

export type EntityResolver = {
  resolveResourceIds: (params: ResolveResourceIdParams) => Promise<string[]>;
  canAccessResource: (params: CanAccessResourceParams) => Promise<boolean>;
  invalidate: (params: InvalidateParams) => Promise<void>;
  init: (models?: string[]) => Promise<void>;
};

export type ScopePath = string[];

export type ResourceDatabase = {
  resolveEntityIds: (
    model: string,
    statements: cardinal.PolicyStatement[],
    entity_filter?: EntityFilter
  ) => Promise<string[]>;
  transact: (model: string, docs: Document[]) => Promise<void>;
  redact: (model: string, ids: string[]) => Promise<void>;
  speculate: (data: SpeculativeData) => Promise<ResourceDatabase>;
};

export type ResourceDatabaseCache = {
  get: (key: string) => Promise<string[] | null>;
  set: (key: string, value: string[]) => Promise<void>;
  invalidate: (keys?: string[]) => Promise<void>;
};

export type ResourceProvider = {
  models: string[];
  fetch: (model: string, ids?: string[]) => Promise<Document[]>;
};
