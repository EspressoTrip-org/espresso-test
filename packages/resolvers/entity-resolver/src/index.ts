export * from './definitions';
export * from './resolver';
export * from './cache';
