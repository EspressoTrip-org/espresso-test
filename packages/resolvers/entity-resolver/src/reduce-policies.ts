import * as cardinal from '@journeyapps-platform/types-cardinal';

export const reduceValidStatements = (policies: cardinal.RawPolicy[], actions: string[], model: string) => {
  return policies.reduce((statements: cardinal.PolicyStatement[], policy) => {
    return policy.statements.reduce((statements, statement) => {
      if (!actions.some((a) => cardinal.compareWithWildcards(statement.actions, a))) {
        return statements;
      }

      if (!statement.resources.some((resource) => cardinal.compareWithWildcards(resource.selector.model, model))) {
        return statements;
      }

      statements.push(statement);

      return statements;
    }, statements);
  }, []);
};
