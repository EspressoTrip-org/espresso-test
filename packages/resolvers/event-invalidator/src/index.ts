export * from './topic-modules';
export * from './invalidator';
