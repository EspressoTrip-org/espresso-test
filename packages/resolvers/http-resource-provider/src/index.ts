import * as entity_resolver from '@journeyapps-platform/recon-entity-resolver';
import * as micro_tracing from '@journeyapps-platform/micro-tracing';
import * as micro_logger from '@journeyapps-platform/micro-logger';
import * as sdk from '@journeyapps-platform/sdk-common';

const logger = micro_logger.createLogger('http-resource-provider');

export type Provider = {
  models: string[];
  endpoint: string;
};

export type ProviderConfig = {
  providers: Provider[];
};

export type CreateHttpResourceProviderParams = ProviderConfig & {
  client?: sdk.NodeNetworkClient;
  headers?: sdk.HeaderCollector;
};

/**
 * Creates an HTTP resource provider for collecting document resources from arbitrary endpoints
 */
export const createHttpResourceProvider = (
  params: CreateHttpResourceProviderParams
): entity_resolver.ResourceProvider => {
  const client = params.client || sdk.createNodeNetworkClient();

  return {
    models: params.providers.map((provider) => provider.models).flat(),
    async fetch(model, ids) {
      return micro_tracing.trace('fetch-resources', async (span) => {
        const provider = params.providers.find((provider) => {
          return provider.models.includes(model);
        });
        if (!provider) {
          throw new Error(`no provider configured for the model '${model}'`);
        }

        span.setAttribute('endpoint', provider.endpoint);

        const res = await client.request(provider.endpoint, {
          headers: params.headers,
          method: sdk.METHOD.POST,
          body: {
            model: model,
            ids: ids
          },
          retryable: true
        });

        /**
         * We want to throw here to ensure the recon DB never gets out of sync.
         *
         * If we ran into 5xx errors then the `retryable: true` option above should handle those. If we burn
         * through all retries before whatever network error is causing the 5xx then further failures should
         * block the consumer that is calling invalidate.
         *
         * If this throws other errors like 4xx errors this this normally indicates configuration errors and
         * we want this to block service boot. In other words the `.init()` will end up throwing
         */
        if (res.response.status >= 300) {
          logger.error(`Failed to collect resources from provider`, {
            status: res.response.status,
            response: await res.response.text()
          });
          throw new Error(`Received status code ${res.response.status}. Could not collect resource`);
        }

        const docs = await res.decode<entity_resolver.Document[]>();
        if (!docs) {
          return [];
        }

        span.setAttributes({
          total_docs: docs.length
        });

        return docs;
      });
    }
  };
};
