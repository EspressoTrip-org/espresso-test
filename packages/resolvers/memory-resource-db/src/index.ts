export * from './referential-clone';
export * from './transactions';
export * from './definitions';
export * from './database';
export * from './queries';
