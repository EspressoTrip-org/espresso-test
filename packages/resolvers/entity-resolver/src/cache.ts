import * as micro_tracing from '@journeyapps-platform/micro-tracing';
import * as defs from './definitions';
import { LRUCache } from 'lru-cache';

// 250MB, assuming it is storing ObjectId's
const DEFAULT_CACHE_SIZE = 250 * 1024 * 1024;

type Params = {
  max?: number;
};
export const createLRUQueryCache = (params?: Params): defs.ResourceDatabaseCache => {
  const cache = new LRUCache<string, string[]>({
    max: params?.max ?? DEFAULT_CACHE_SIZE,
    sizeCalculation(value) {
      return value.length * 24;
    }
  });

  return {
    async get(key) {
      return micro_tracing.trace('cache-get', () => {
        return cache.get(key) || null;
      });
    },
    async set(key, ids) {
      micro_tracing.trace('cache-set', () => {
        cache.set(key, ids);
      });
    },
    async invalidate(maybe_keys) {
      micro_tracing.trace('cache-invalidate', () => {
        let keys = maybe_keys;
        if (!keys || keys.length === 0) {
          keys = Array.from(cache.keys());
        }

        for (const key of keys) {
          cache.delete(key);
        }
      });
    }
  };
};
