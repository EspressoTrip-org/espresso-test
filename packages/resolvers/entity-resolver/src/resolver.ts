import * as micro_tracing from '@journeyapps-platform/micro-tracing';
import * as micro_logger from '@journeyapps-platform/micro-logger';
import * as reducer from './reduce-policies';
import * as defs from './definitions';

const logger = micro_logger.createLogger('recon-entity-resolver');

const inverseModelMappings = (providers: defs.ResourceProvider[]) => {
  return providers.reduce((mapping: Record<string, defs.ResourceProvider>, provider) => {
    return provider.models.reduce((mapping, model) => {
      mapping[model] = provider;
      return mapping;
    }, mapping);
  }, {});
};

export type CreateEntityResolverParams = {
  db: defs.ResourceDatabase;
  resource_providers: defs.ResourceProvider[];
  default_scope_path?: defs.ScopePath;
};

export const createEntityResolver = (params: CreateEntityResolverParams): defs.EntityResolver => {
  const { db } = params;

  const model_mapping = inverseModelMappings(params.resource_providers);

  const fetch = async (model: string, ids?: string[]) => {
    const provider = model_mapping[model];

    if (!provider) {
      logger.warn('no providers serve the requested model');
      return [];
    }

    return provider.fetch(model, ids);
  };

  return {
    async resolveResourceIds(params) {
      const speculative_keys = Object.keys(params.speculative_data || {});

      return micro_tracing.trace('resolve-entity-ids', async (span) => {
        span.setAttributes({
          actions: params.actions,
          model: params.model,
          speculative_keys: speculative_keys,
          speculative: speculative_keys.length > 0
        });

        if (params.entity_filter) {
          span.addEvent('entity-filter', params.entity_filter);
        }

        const statements = reducer.reduceValidStatements(params.policies, params.actions || ['read'], params.model);

        // Produce a new speculative DB from an injected data set. This will not be persisted to the
        // original DB
        let query_db = db;
        if (speculative_keys.length > 0) {
          query_db = await query_db.speculate(params.speculative_data || {});
        }

        const ids = await query_db.resolveEntityIds(params.model, statements, params.entity_filter);
        span.setAttribute('resolved-ids', ids.length);
        return ids;
      });
    },

    async canAccessResource(params) {
      const ids = await this.resolveResourceIds({
        ...params,
        entity_filter: {
          id: params.id
        }
      });
      return ids.length > 0;
    },

    async invalidate(params) {
      await micro_tracing.trace('invalidate', async (span) => {
        span.setAttributes({
          model: params.model,
          ids: params.ids
        });

        const documents = await fetch(params.model, params.ids);
        span.setAttribute('documents', documents.length);

        if (params.ids) {
          const redaction_ids = params.ids.filter((id) => {
            return !!documents.find((doc) => {
              return doc.id === id;
            });
          });
          if (redaction_ids.length) {
            await db.redact(params.model, redaction_ids);
          }
        }

        await db.transact(params.model, documents);
      });
    },
    async init(custom_models) {
      const models = custom_models || Object.keys(model_mapping);
      await micro_tracing.trace('init', async (span) => {
        span.setAttribute('models', models);

        for (const model of models) {
          logger.info(`initializing ${model}`);

          const documents = await fetch(model);
          span.setAttribute(`${model}.documents`, documents.length);

          await db.transact(model, documents);
        }
      });
    }
  };
};
