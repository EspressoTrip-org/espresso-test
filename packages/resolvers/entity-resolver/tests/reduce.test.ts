import { describe, it, expect } from 'vitest';

import { reduceValidStatements } from '../src/reduce-policies';
import * as types from '@journeyapps-platform/types-cardinal';

describe('statement-filtering', () => {
  const fixtures: types.RawPolicy[] = [
    {
      statements: [
        {
          actions: ['a', 'b'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'waffle',
                id: '*'
              }
            }
          ]
        },
        {
          actions: ['c'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'waffle',
                id: '*'
              }
            }
          ]
        }
      ]
    },
    {
      statements: [
        {
          actions: ['d'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'waffle',
                id: '*'
              }
            }
          ]
        }
      ]
    },
    {
      statements: [
        {
          actions: ['d'],
          resources: [
            {
              scope: '123',
              selector: {
                model: 'not-waffle',
                id: '*'
              }
            }
          ]
        }
      ]
    }
  ];

  it('should handle a single element actions array', () => {
    expect(reduceValidStatements(fixtures, ['a'], 'waffle')).toMatchSnapshot();
    expect(reduceValidStatements(fixtures, ['b'], 'waffle')).toMatchSnapshot();
    expect(reduceValidStatements(fixtures, ['d'], 'waffle')).toMatchSnapshot();
  });

  it("should handle multiple actions with 'or'", () => {
    expect(reduceValidStatements(fixtures, ['a', 'd'], 'waffle')).toMatchSnapshot();
  });
});
