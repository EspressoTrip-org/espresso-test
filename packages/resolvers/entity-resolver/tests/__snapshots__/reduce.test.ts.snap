// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`statement-filtering > should handle a single element actions array 1`] = `
[
  {
    "actions": [
      "a",
      "b",
    ],
    "resources": [
      {
        "scope": "123",
        "selector": {
          "id": "*",
          "model": "waffle",
        },
      },
    ],
  },
]
`;

exports[`statement-filtering > should handle a single element actions array 2`] = `
[
  {
    "actions": [
      "a",
      "b",
    ],
    "resources": [
      {
        "scope": "123",
        "selector": {
          "id": "*",
          "model": "waffle",
        },
      },
    ],
  },
]
`;

exports[`statement-filtering > should handle a single element actions array 3`] = `
[
  {
    "actions": [
      "d",
    ],
    "resources": [
      {
        "scope": "123",
        "selector": {
          "id": "*",
          "model": "waffle",
        },
      },
    ],
  },
]
`;

exports[`statement-filtering > should handle multiple actions with 'or' 1`] = `
[
  {
    "actions": [
      "a",
      "b",
    ],
    "resources": [
      {
        "scope": "123",
        "selector": {
          "id": "*",
          "model": "waffle",
        },
      },
    ],
  },
  {
    "actions": [
      "d",
    ],
    "resources": [
      {
        "scope": "123",
        "selector": {
          "id": "*",
          "model": "waffle",
        },
      },
    ],
  },
]
`;
